flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricFetcherTest.java

```diff
@@ -263,7 +220,151 @@ public void testShortUpdateInterval() throws InterruptedException {
+            JobID jobID,
+            ResourceID tmRID,
+            Time timeout,
+            long updateInterval,
+            long waitTimeBeforeReturnMetricResults,
+            @Nullable AtomicInteger requestMetricQueryServiceGatewaysCounter) {
+        final ExecutorService executor = java.util.concurrent.Executors.newSingleThreadExecutor();
+        // ========= setup QueryServices
+        // ================================================================================
+
+        final MetricQueryServiceGateway jmQueryService =
+                new TestingMetricQueryServiceGateway.Builder()
+                        .setQueryMetricsSupplier(
+                                () ->
+                                        CompletableFuture.completedFuture(
+                                                new MetricDumpSerialization
+                                                        .MetricSerializationResult(
+                                                        new byte[0],
+                                                        new byte[0],
+                                                        new byte[0],
+                                                        new byte[0],
+                                                        0,
+                                                        0,
+                                                        0,
+                                                        0)))
+                        .build();
+
+        MetricDumpSerialization.MetricSerializationResult requestMetricsAnswer =
+                createRequestDumpAnswer(tmRID, jobID);
+        final MetricQueryServiceGateway tmQueryService =
+                new TestingMetricQueryServiceGateway.Builder()
+                        .setQueryMetricsSupplier(
+                                () -> {
+                                    if (waitTimeBeforeReturnMetricResults > 0) {
+                                        CompletableFuture<
+                                                        MetricDumpSerialization
+                                                                .MetricSerializationResult>
+                                                metricsAnswerFuture = new CompletableFuture<>();
+                                        CompletableFuture.completedFuture(null)
+                                                .thenRunAsync(
+                                                        waitTimeMs(
+                                                                waitTimeBeforeReturnMetricResults),
+                                                        executor)
+                                                .whenCompleteAsync(
+                                                        (ignore, throwable) -> {
+                                                            if (throwable != null) {
+                                                                fail(throwable.getMessage());
```

Comment:
We should try to avoid using the `fail` method. Maybe we can use other approach to do the similar logic, such as making the future complete with exception.